tangelo/toolboxes/circuits/tests/test_lcu.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2021 Good Chemistry Company.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+import math
+
+from openfermion import get_sparse_operator
+import numpy as np
+from scipy.linalg import expm
+
+from tangelo.linq import Simulator, backend_info
+from tangelo.helpers.utils import installed_backends
+from tangelo.linq.helpers.circuits.statevector import StateVector
+from tangelo.toolboxes.operators.operators import QubitOperator
+from tangelo.toolboxes.qubit_mappings.mapping_transform import fermion_to_qubit_mapping
+from tangelo.toolboxes.ansatz_generator.ansatz_utils import get_qft_circuit
+from tangelo.molecule_library import mol_H2_sto3g
+from tangelo.toolboxes.circuits.lcu import get_lcu_circuit, get_truncated_taylor_series
+
+# Test for both "cirq" and if available "qulacs". These have different orderings.
+# qiskit is not currently supported because does not have multi controlled general gates.
+backends = ["cirq", "qulacs"] if "qulacs" in installed_backends else ["cirq"]
+# Initiate Simulator using cirq for phase estimation tests as it has the same ordering as openfermion
+# and we are using an exact eigenvector for testing.
+sim_cirq = Simulator("cirq")
+
+
+class lcu_Test(unittest.TestCase):
+
+    def test_get_truncated_taylor_series(self):
+        """Test time-evolution of truncated Taylor series for different orders and times"""
+
+        qu_op = fermion_to_qubit_mapping(mol_H2_sto3g.fermionic_hamiltonian, "scbk", mol_H2_sto3g.n_active_sos, mol_H2_sto3g.n_active_electrons,
+                                         True, 0)
+        n_qubits_qu_op = math.ceil(math.log2(len(qu_op.terms)))
+        ham = get_sparse_operator(qu_op).toarray()
+        _, vecs = np.linalg.eigh(ham)
+        vec = (vecs[:, 0] + vecs[:, 1])/np.sqrt(2)
+
+        time = 1.9
+        exact = expm(-1j*ham*time)@vec
+
+        for backend in backends:
+            sim = Simulator(backend)
+            statevector_order = backend_info[backend]["statevector_order"]
+            sv = StateVector(vec, order=statevector_order)
+            sv_circuit = sv.initializing_circuit()
+
+            # Tested for up to k = 5 but 5 is very slow.
```

Comment:
It would be good to know why it's slow (e.g what part of the computation scales "badly" with k).
You could very quickly print 2-3 timers that use functions you think are responsible to get that info.

Maybe the answer is simply that simulators are just spending a lot of time simulating circuits.

Reply:
The main reason that the algorithm slows done for each added k is that the number of qubits increases by 4 for each additional k. So k=5 has 23 qubits compared to 19 qubits for k=4. 

The scaling of the gates was also exponential but your question made me realize that I had an inefficiency in the code. I now added a `get_unary_prep` function that does the unary encoding in linear gates so the test with `k=5` takes 30s using qulacs. I don't think n=5 is necessary for testing purposes.

Reply:
awesome, glad the comment was useful somehow.